servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,661 @@
+                                    FlowControlUtils::addWithOverflowProtection);
+                        }
+                    }
+                } catch (Throwable cause) {
+                    delayedCause = catchUnexpected(delayedCause, cause);
+                } finally {
+                    tryAcquire = !releaseLock(requestingLockUpdater, this);
+                }
+            }
+            if (delayedCause != null) {
+                throwException(delayedCause);
+            }
+        }
+
+        private int calculateRequestNQuota(long availableRequestN) {
+            // Get an approximate quota to distribute to each active mapped subscriber. If
+            // activeMappedSources changes (subscriber added or terminated) we will re-distribute.
+            final int prevActiveSources = activeMappedSources;
+            return (int) min(Integer.MAX_VALUE,
+                    max(prevActiveSources > 0 ? availableRequestN / prevActiveSources : availableRequestN,
+                            source.minMappedDemand));
+        }
+
+        private void doCancel(boolean cancelSubscription) {
+            // Prevent future onNext operations from adding to subscribers which otherwise may result in
+            // not cancelling mapped Subscriptions. This should be Integer.MIN_VALUE to prevent subsequent mapped
+            // source completion from incrementing the count to 0 or positive as terminateActiveMappedSources flips
+            // the count to negative to signal to mapped sources we have completed.
+            activeMappedSources = Integer.MIN_VALUE;
+            pendingDemand = -1;
+            try {
+                if (cancelSubscription) {
+                    assert subscription != null;
+                    subscription.cancel();
+                }
+            } finally {
+                cancellableSubscribers.cancel();
+                // Don't bother clearing out hungrySubscribers or signals (which require additional concurrency control)
+                // because it is assumed this Subscriber will be dereferenced and eligible for GC [1].
+                // [1] https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.3/README.md#3.13
+            }
+        }
+
+        private boolean tryDecrementPendingDemand() {
+            for (;;) {
+                final long prevDemand = pendingDemand;
+                if (prevDemand <= 0) {
+                    return false;
+                } else if (pendingDemandUpdater.compareAndSet(this, prevDemand, prevDemand - 1)) {
+                    return true;
+                }
+            }
+        }
+
+        private void tryEmitItem(Object item, FlatMapPublisherSubscriber<T, R> subscriber) {
+            if (tryAcquireLock(emittingLockUpdater, this)) { // fast path. no concurrency, try to skip the queue.
+                try {
+                    if (subscriber.hasSignalsQueued() || (needsDemand(item) && !tryDecrementPendingDemand())) {
```

Comment:
No, `signalsQueued` is to avoid out of order delivery of signals relative to each mapped publisher. Either we go through the queue for all signals and ordering is provided by the queue, or if we may skip the queue (like we are doing now) we need to check if there maybe items already queued (in which case we need to go through the queue to preserve ordering). The `signalsQueued` flag is currently cleared after demand is exhausted as we anyways need a reference to the `FlatMapPublisherSubscriber` to request more demand. We could potentially clear this flag earlier but we would need a reference to the `FlatMapPublisherSubscriber` while processing `onNext` which would require object allocation.